File: apps/api/src/reception/quality-validation.service.ts
import { Injectable } from "@nestjs/common";
import { QualityParameter, TransactionStatus } from "@cc-mc/shared-types";

export interface QualityReadingInput {
  fat: number;
  snf: number;
  temperature: number;
}

export interface ResolvedRule {
  minValue: number;
  maxValue: number;
}

export interface QualityValidationResult {
  status: TransactionStatus.ACCEPTED | TransactionStatus.HOLD;
  reason: string | null;
}

/**
 * Isolated quality-validation business logic (kept out of the controller
 * and out of ReceptionService's persistence code, per the project's
 * "Quality Validation" instruction: `validateQuality(reading, rules) ->
 * ACCEPT/REJECT/HOLD`).
 *
 * MVP interpretation of BRD Sections 13 & 38 (documented in
 * docs/assumptions.md #auto-reject-vs-hold): the BRD's own diagram shows
 * an out-of-limit reading going to HOLD, then a Manager resolving it to
 * ACCEPT or REJECT. Nothing in the BRD describes the system ever
 * auto-rejecting on its own. So this function only ever returns ACCEPTED
 * or HOLD - REJECTED is reachable exclusively through a Manager override
 * of a HOLD (see ReceptionService.override). This is a documented
 * assumption, not an invented protocol detail.
 */
@Injectable()
export class QualityValidationService {
  validate(reading: QualityReadingInput, rules: Map<QualityParameter, ResolvedRule>): QualityValidationResult {
    const failures: string[] = [];

    const checks: Array<[QualityParameter, number]> = [
      [QualityParameter.FAT, reading.fat],
      [QualityParameter.SNF, reading.snf],
      [QualityParameter.TEMPERATURE, reading.temperature],
    ];

    for (const [param, value] of checks) {
      const rule = rules.get(param);
      if (!rule) {
        // ReceptionService is responsible for resolving all required rules
        // before calling validate(); a missing rule here is a programming
        // error, not a business outcome.
        throw new Error(`No resolved rule provided for parameter ${param}`);
      }
      if (value < rule.minValue || value > rule.maxValue) {
        failures.push(`${param} (${value}) outside configured range [${rule.minValue}, ${rule.maxValue}]`);
      }
    }

    if (failures.length === 0) {
      return { status: TransactionStatus.ACCEPTED, reason: null };
    }

    return {
      status: TransactionStatus.HOLD,
      reason: `Quality outside limits: ${failures.join("; ")}`,
    };
  }
}

File: src/CCMC.Cloud.Domain/Services/QualityValidationService.cs
using CCMC.Cloud.Domain.Enums;

namespace CCMC.Cloud.Domain.Services;

public sealed record QualityReadingInput(decimal Fat, decimal Snf, decimal Temperature);

public sealed record ResolvedQualityRule(decimal MinValue, decimal MaxValue);

public sealed record QualityValidationResult(TransactionStatus Status, string? Reason)
{
    public static QualityValidationResult Accepted() => new(TransactionStatus.Accepted, null);
    public static QualityValidationResult Hold(string reason) => new(TransactionStatus.Hold, reason);
}

/// <summary>
/// The server's authoritative quality validation - the cloud is the final
/// word on ACCEPTED/HOLD, never merely trusting a client-side evaluation
/// (this session's explicit instruction: "Do not automatically reject milk
/// solely based on client-side quality rule evaluation"). Deliberately kept
/// identical in behavior to the Windows client's own QualityValidationService
/// (independently implemented here, not shared code) so both sides reach the
/// same decision for the same inputs: out-of-range readings only ever
/// produce HOLD, never an automatic REJECTED - REJECTED is reachable only
/// through a Manager's explicit override of a HOLD.
/// </summary>
public static class QualityValidationService
{
    public static QualityValidationResult Validate(
        QualityReadingInput reading,
        IReadOnlyDictionary<QualityParameter, ResolvedQualityRule> rules)
    {
        var failures = new List<string>();

        void Check(QualityParameter parameter, decimal value)
        {
            if (!rules.TryGetValue(parameter, out var rule))
            {
                throw new InvalidOperationException(
                    $"No resolved quality rule supplied for parameter {parameter}. " +
                    "Callers must resolve every required rule before calling Validate().");
            }

            if (value < rule.MinValue || value > rule.MaxValue)
            {
                failures.Add($"{parameter} ({value}) outside configured range [{rule.MinValue}, {rule.MaxValue}]");
            }
        }

        Check(QualityParameter.Fat, reading.Fat);
        Check(QualityParameter.Snf, reading.Snf);
        Check(QualityParameter.Temperature, reading.Temperature);

        return failures.Count == 0
            ? QualityValidationResult.Accepted()
            : QualityValidationResult.Hold($"Quality outside limits: {string.Join("; ", failures)}");
    }
}
